enter.c

```diff
@@ -225,13 +226,65 @@ int enter(struct entry_settings *opts)
 		}
 	}
 
+	int rootpid = getpid();
+	bool req_cgroup_clean = false;
+	int cleanfd;
+
+	// Do not join the root cgroup if limits are to be applied, otherwise bst needs to create sub-hierarchies
+	if (!opts->nactiveclimits && opts->cgroup.present) {
+		int cgroupfd = open(opts->cgroup.cgroup_path, O_PATH | O_DIRECTORY);
+		if (cgroupfd == -1) {
+			err(1, "unable open cgroup %s", opts->cgroup.cgroup_path);
+		}
+
+		join_cgroup(opts->cgroup.cgroup_path, cgroupfd);
+
+		close(cgroupfd);
+
+	} else if (opts->nactiveclimits && opts->cgroup.present) {
+
+		int rootfd = open(root, O_PATH, 0);
+		if (rootfd == -1) {
+			 err(1, "open %s", root);
+		}
+
+		const char * rootcgroup = opts->cgroup.cgroup_path;
+		const char * subcgroup = makepath("%s/bst.%d", rootcgroup,rootpid);
+
+		if (mkdir(subcgroup, 0755) != -1) {
+
+			// We are able to create new cgroups, and there are limits to apply
+			cleanfd = open(rootcgroup, O_PATH | O_DIRECTORY);
+			if (cleanfd == -1) {
+				err(1, "unable open cgroup %s", rootcgroup);
+			}
+
+			apply_climits(opts->climits, subcgroup, cleanfd);
+
+			int cgroupfd = open(subcgroup, O_PATH | O_DIRECTORY);
+			if (cgroupfd == -1) {
+				err(1, "unable open cgroup %s", subcgroup);
+			}
+
+			join_cgroup(subcgroup, cgroupfd);
+			close(cgroupfd);
+
+			// The creation of this sub-hierarchy will require later cleanup
+			req_cgroup_clean = true;
```

Comment:
How about initializing cleanfd to -1, and use this as a condition for determining whether it needs cleanup or not?

Reply:
Good thought.